src/expand.rs

```diff
@@ -95,9 +102,236 @@ fn impl_struct(input: &DeriveInput, data: &DataStruct) -> Result<TokenStream> {
+                    TypeParamBound::Trait(trait_bound) => Some(trait_bound),
+                    _ => None,
+                })
+                .cloned()
+                .collect();
+            (type_param.ident.clone(), trait_bounds)
+        })
+        .collect();
+
+    for predicate in where_clause.predicates.iter() {
+        match predicate {
+            WherePredicate::Type(ref pred_ty) => {
+                let known_bounds: &mut Vec<TraitBound> = match &pred_ty.bounded_ty {
+                    Type::Path(TypePath { path, qself: None }) => {
+                        assert!(
+                            path.leading_colon.is_none(),
+                            "LHS of `where` boundary can't resolve from global scope!"
+                        );
+                        assert_eq!(
+                            1,
+                            path.segments.len(),
+                            "LHS of `where` boundary can't have segments!"
+                        );
+                        let cur_segment = path.segments.first().unwrap();
+                        assert_eq!(PathArguments::None, cur_segment.arguments);
+                        let (_, ref mut known_bounds) = param_constraint_mapping
+                            .iter_mut()
+                            .find(|(id, _)| *id == cur_segment.ident)
+                            .expect("expected all types in `where` boundaries to be declared!");
+                        known_bounds
+                    }
+                    x => unreachable!("expected simple type in where clause, but got {:?}", x),
+                };
+                for bound in pred_ty.bounds.iter() {
+                    match bound {
+                        TypeParamBound::Trait(ref bound) => {
+                            known_bounds.push(bound.clone());
+                        }
+                        _ => (),
+                    }
+                }
+            }
+            _ => (),
+        }
+    }
+
+    param_constraint_mapping
+}
+
+/// Hygienically add `where _: Display` to the set of [TypeParamBound]s for `ident`, creating such
+/// a set if necessary.
+fn ensure_display_in_where_clause_for_type(where_clause: &mut WhereClause, ident: Ident) {
+    // Option<()> is a contortion to avoid multiple mutable references to `where_clause`.
+    let maybe_constrained: Option<()> = where_clause
+        .predicates
+        .iter_mut()
+        // Find the `where` predicate constraining the current type param, if it exists.
+        .flat_map(|predicate| match predicate {
+            WherePredicate::Type(pred_ty) => Some(pred_ty),
+            // We're looking through type constraints, not lifetime constraints.
+            _ => None,
+        })
+        .find(|pred_ty| match &pred_ty.bounded_ty {
+            Type::Path(TypePath { path, .. }) => ident == path.segments.first().unwrap().ident,
+            x => unreachable!("expected simple type in where clause, but got {:?}", x),
```

Comment:
same as above regarding debug dumping syn data structures

Reply:
I have removed this panic as well!